src/Microsoft.AspNetCore.Mvc.Core/ServiceFilterAttribute.cs

```diff
@@ -37,6 +37,20 @@ public ServiceFilterAttribute(Type type)
             }
 
             ServiceType = type;
+
+            if (typeof(IFilterFactory).IsAssignableFrom(ServiceType))
+            {
+                throw new InvalidOperationException(
+                    Resources.FormatCannotUseFilterFactoryWithinFilterFactory(ServiceType, GetType()));
+            }
+
+            if (!typeof(IFilterMetadata).IsAssignableFrom(ServiceType))
```

Comment:
This is a slight breaking behavior change:

```C#
interface IService {} 
    
public class MyType : IService, IFilterMetadata {}

services.AddSingleton<IService, MyType>();
    
[ServiceFilter(typeof(IService))]
```

Previously this would have worked just fine, versus an exception thrown now. Could we leave this check in the `CreateInstance` call?

Reply:
Good point. I think the changes here are not worth taking the risk in the end. @rynowak mentioned about https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Authorization/AuthorizeFilter.cs#L24 which is both a filter and filterfactory and would be broken with my changes. Users could also have this type of code.